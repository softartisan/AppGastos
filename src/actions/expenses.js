import uuid from 'uuid';
import database from '../firebase/firebase';

//ADD_EXPENSE
const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
    description = '',
    note ='',
    amount = 0,
    createdAt = 0 
    } = expenseData;

    const expense = {
      description,
      note,
      amount,
      createdAt
    }

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id:ref.key,
        ...expense
      }));
    });

  }
}



//REMOVE_EXPENSE
const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    idToRemove: id
});

const startRemoveExpense = ({id} = {}) => {

  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}));
    })
  }

}


//EDIT_EXPENSE
const editExpense = (id,updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

const startEditExpense = (id, updates) => {
// database.ref().update({
//   stressLevel: 9,
//   'job/company': 'Amazon',
//   'location/city': 'Seattle'
// });
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update({
      ...updates
    }).then(() => {
      dispatch(editExpense(id, updates));
    })
  }

}

//SET_EXPENSES
const setExpenses = (expenses) => {
  return { type: 'SET_EXPENSES', expenses }
}

const startSetExpenses = () => {
  //Retorno una funciÃ³n
  return (dispatch) => {
    //Saco expenses desde firebase y las guardo en expenses
    //Retorno esta Promise
    return database.ref('expenses').once('value').then((snapshot) => {

      const expenses = [];
      //Por cada objeto de firebase lleno un objeto del array expenses
      snapshot.forEach((childSnapshot) => {
        expenses.push( {id: childSnapshot.key, ...childSnapshot.val()} );
      });

      dispatch(setExpenses(expenses));

    });
  }
}

export {
    startEditExpense,
    editExpense,
    removeExpense,
    startRemoveExpense,
    addExpense,
    startAddExpense,
    setExpenses,
    startSetExpenses
}