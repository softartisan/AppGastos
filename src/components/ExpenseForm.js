import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

const now = moment();

class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: '',
            isAdding: props.expense ? 'Editar Gasto' : 'Agregar Gasto'
        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ({amount}));
        }
    }
    onDateChange = (createdAt) => {
        (createdAt) && this.setState(() => ({createdAt}));
        
    }
    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}));
    }
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(() => ({error: 'Ingresa un monto y descripción valida.'}));
        }else{
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }
    render() {
        return (
            <div>
                <form className="form" onSubmit={this.onSubmit}>
                    {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <input
                        type='text'
                        placeholder="Descripción"
                        className="text-input"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type='text'
                        className="text-input"
                        value={this.state.amount}
                        placeholder="Monto (CLP)"
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        className="textarea"
                        placeholder="Agrega una nota (opcional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                   <div>
                   <button className="button">{this.state.isAdding}</button>
                   </div>
                </form>
            </div>
        );
    }
}

export {
    ExpenseForm as default
}