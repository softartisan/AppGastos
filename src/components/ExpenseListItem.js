import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({dispatch, id, description, amount, createdAt}) => (
    <div>
        <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
        </Link>
        <p>
        {numeral(amount).format('$000,000,000')} 
        - 
        {moment(createdAt).format('DD MMMM YYYY')}
        </p>
    </div>
);

export default ExpenseListItem;