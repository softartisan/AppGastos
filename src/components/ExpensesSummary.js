import React from 'react';
import {connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = (props) => {
    const formattedExpenseTotal = numeral(props.expensesTotal).format('$0,000');
    const expenseWord = props.expenseCount===1 ? 'gasto' : 'gastos';
  return (
    <div className="page-header">
      <div className='content-container'>
        <h1 className='page-header__title'>Viendo <span>{props.expenseCount}</span> {expenseWord} con un total de <span>{formattedExpenseTotal}</span></h1>
        <div className='page-header__actions'>
          <Link className="button" to='/create'>Nuevo Gasto</Link>
        </div>
      </div>
    </div>
  );
}


const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expensesTotal: getExpensesTotal(visibleExpenses),
    expenseCount: visibleExpenses.length
  }
}

export default connect(mapStateToProps)(ExpensesSummary);