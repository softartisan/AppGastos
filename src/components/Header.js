import React from 'react';
import {NavLink} from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>Inicio</NavLink>
        <NavLink to="/create" activeClassName="is-active">Crear</NavLink>
        
    </header>
);

export default Header;