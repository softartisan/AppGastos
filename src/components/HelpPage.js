import React from 'react';

const HelpPage = () => (
    <div>
        Esto es desde el Help
    </div>
);

export default HelpPage;