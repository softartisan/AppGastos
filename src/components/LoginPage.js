import React from 'react';
import { connect } from 'react-redux';
import {startLogin} from '../actions/auth';

export const LoginPage = ({startLogin}) => (
  <div className="box-layout">
  <div className='box-layout__box'>
    <h1 className="box-layout__title">Gastos App</h1>
    <p>Es tiempo de que controles tus gastos.</p>
    <button className='button' onClick={startLogin}>Iniciar sesi√≥n con Google</button>
  </div>
  </div>
)

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
})

export default connect(undefined, mapDispatchToProps)(LoginPage);
