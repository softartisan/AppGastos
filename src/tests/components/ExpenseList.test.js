import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
import {ExpenseList} from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('should render expenselist correctly',()=>{
  const wrapper = shallow(<ExpenseList expenses={expenses}/>);
  expect(wrapper).toMatchSnapshot();
});

test('Should render ExpenseList with empty message', () => {
  const wrapper = shallow(<ExpenseList expenses={[]}/>);
  expect(wrapper).toMatchSnapshot();
});