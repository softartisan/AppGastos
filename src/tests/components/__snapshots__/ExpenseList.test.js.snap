// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseList with empty message 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ExpenseList
    expenses={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        No expenses
      </p>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "No expenses",
      },
      "ref": null,
      "rendered": "No expenses",
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          No expenses
        </p>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "No expenses",
        },
        "ref": null,
        "rendered": "No expenses",
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`should render expenselist correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ExpenseList
    expenses={
      Array [
        Object {
          "amount": 100,
          "createdAt": 0,
          "description": "cafe",
          "id": "1",
          "note": "",
        },
        Object {
          "amount": 150,
          "createdAt": -345600000,
          "description": "atun",
          "id": "2",
          "note": "",
        },
        Object {
          "amount": 4500,
          "createdAt": 345600000,
          "description": "credit card",
          "id": "3",
          "note": "",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ExpenseListItem
          amount={100}
          createdAt={0}
          description="cafe"
          id="1"
          note=""
        />,
        <ExpenseListItem
          amount={150}
          createdAt={-345600000}
          description="atun"
          id="2"
          note=""
        />,
        <ExpenseListItem
          amount={4500}
          createdAt={345600000}
          description="credit card"
          id="3"
          note=""
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "amount": 100,
          "createdAt": 0,
          "description": "cafe",
          "id": "1",
          "note": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "function",
        "props": Object {
          "amount": 150,
          "createdAt": -345600000,
          "description": "atun",
          "id": "2",
          "note": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "function",
        "props": Object {
          "amount": 4500,
          "createdAt": 345600000,
          "description": "credit card",
          "id": "3",
          "note": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ExpenseListItem
            amount={100}
            createdAt={0}
            description="cafe"
            id="1"
            note=""
          />,
          <ExpenseListItem
            amount={150}
            createdAt={-345600000}
            description="atun"
            id="2"
            note=""
          />,
          <ExpenseListItem
            amount={4500}
            createdAt={345600000}
            description="credit card"
            id="3"
            note=""
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "amount": 100,
            "createdAt": 0,
            "description": "cafe",
            "id": "1",
            "note": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "amount": 150,
            "createdAt": -345600000,
            "description": "atun",
            "id": "2",
            "note": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "function",
          "props": Object {
            "amount": 4500,
            "createdAt": 345600000,
            "description": "credit card",
            "id": "3",
            "note": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
